### PR TITLE
Isaicp 2579

### PR DESCRIPTION
**Pull request by brummbar**
By submitting a pull request to the Joinup repository:
- you accept the conditions in the Joinup Individual Contributor Assignment Agreement
- you accept the terms laid out in the the code of conduct.

### CODE REVIEW

File: tests/src/Context/DrupalContext.php [hunk 1]
```diff
@@ -219,4 +219,20 @@ public function assertFieldsetsPresent($fieldsets) {
...
   }
 
+  /**
+   * Click on an element by css class.
+   *
+   * @Then /^I click on element "([^"]*)"$/
+   */
+  public function iClickOn($element) {
+    $page = $this->getSession()->getPage();
+    $findName = $page->find("css", $element);
```
Comment (brummbar) [inline tests/src/Context/DrupalContext.php]: Use single quotes here.

File: web/themes/joinup/joinup_theme.theme [hunk 2]
```diff
@@ -8,26 +8,37 @@
...
+  // Set page title.
   $title = \Drupal::service('title_resolver')->getTitle($request, $route_match->getRouteObject());
-  $variables['title'] = $title;
 
   $route_name = $route_match->getRouteName();
   $blue_header_routes = array(
     'user.login',
     'user.register',
     'user.pass',
+    'node.add',
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: What about sorting this alphabetically? @pfrenssen suggestion eheh

File: web/themes/joinup/joinup_theme.theme [hunk 3]
```diff
@@ -8,26 +8,37 @@
...
 
 /**
  * Implements hook_preprocess_page().
  */
 function joinup_theme_preprocess_page(&$variables) {
-  // Set page title.
   $request = \Drupal::request();
   $route_match = \Drupal::routeMatch();
+
+  // Set page title.
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: Documentation here could be improved, saying what we are trying to do with the code here. For example
`// We need to show the header in a blue variant on certain routes`

File: web/themes/joinup/joinup_theme.theme [hunk 4]
```diff
@@ -8,26 +8,37 @@
...
+    'rdf_entity.rdf_add',
+    'entity.node.edit_form',
+    'collection.propose_form',
+    'rdf_entity.rdf_add_page',
+    'entity.rdf_entity.edit_form',
   );
 
   if (!empty($route_name) && in_array($route_name, $blue_header_routes)) {
     $variables['is_blue_header'] = TRUE;
+    $title = $variables['page']['content']['pagetitle'];
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: Again here, something like
`// Move the title from the content region to {...}`.

File: web/themes/joinup/joinup_theme.theme [hunk 5]
```diff
@@ -43,6 +54,34 @@ function joinup_theme_preprocess_page(&$variables) {
...
+}
+
+/**
+ * Implements hook_preprocess_page_title().
+ */
+function joinup_theme_preprocess_page_title(&$variables) {
+  $route_match = \Drupal::routeMatch();
+  $route_name = $route_match->getRouteName();
+
+  // Array contains create pages.
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: Here also improve the documentation. Something explaining the reason of what we are doing.
`// Set a flag when we are on content creation routes in order to customise some templates.`

File: web/themes/joinup/joinup_theme.theme [hunk 6]
```diff
@@ -116,23 +142,70 @@ function joinup_theme_preprocess_fieldset(&$variables) {
...
+  $route_name = $route_match->getRouteName();
+  $form_routes = array(
+    'node.add',
+    'rdf_entity.rdf_add',
+    'entity.node.edit_form',
+    'collection.propose_form',
+    'entity.rdf_entity.edit_form',
+  );
+
+  // Check if container is field element - it has only one parent.
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: Why is this needed? In which case we have more than one parent?

File: web/themes/joinup/joinup_theme.theme [hunk 7]
```diff
@@ -116,23 +142,70 @@ function joinup_theme_preprocess_fieldset(&$variables) {
...
+    $variables['is_form_wrapper'] = TRUE;
+  }
+}
+
+/**
  * Implements hook_theme_suggestions_alter().
  */
 function joinup_theme_theme_suggestions_alter(array &$suggestions, array $variables) {
-  if (isset($variables['element']) && isset($variables['element']['#type'])) {
+  if (isset($variables['element']['#type'])) {
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: You can swap the `isset()` with `!empty()` as it will check the presence of each key.

File: web/themes/joinup/joinup_theme.theme [hunk 8]
```diff
@@ -116,23 +142,70 @@ function joinup_theme_preprocess_fieldset(&$variables) {
...
+    && count($variables['element']['#parents']) == 1
+    && $variables['element']['#id'] != 'edit-actions') {
+    $variables['is_form_wrapper'] = TRUE;
+  }
+}
+
+/**
  * Implements hook_theme_suggestions_alter().
  */
 function joinup_theme_theme_suggestions_alter(array &$suggestions, array $variables) {
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: You can provide a third parameter here, `$hook`, and replace the `$variables['theme_hook_original']` usage.

File: web/themes/joinup/joinup_theme.theme [hunk 9]
```diff
@@ -116,23 +142,70 @@ function joinup_theme_preprocess_fieldset(&$variables) {
...
+      'textfield',
+      'url',
+      'password',
+      'email',
+      'entity_autocomplete',
+    );
+
     if ($variables['element']['#type'] == 'form') {
-      $original_theme_hook = $variables['theme_hook_original'];
       $suggestions[] = $original_theme_hook . '__' . str_replace('-', '_', $variables['element']['#id']);
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: There is an issue here. The html id is being used, but Drupal appends a counter to it in some cases (ajax, multiple forms in same page). The safe string to use is '#form_id', which the actually id is based on (e.g. 'rdf_entity_collection_propose_form').

File: web/themes/joinup/joinup_theme.theme [hunk 10]
```diff
@@ -116,23 +142,70 @@ function joinup_theme_preprocess_fieldset(&$variables) {
...
       $suggestions[] = $original_theme_hook . '__' . str_replace('-', '_', $variables['element']['#type']);
     }
 
     $slider_ids = array(
       'edit-field-ar-elibrary-creation',
-      'edit-field-test-lista',
-      'edit-field-test-luist-normal',
     );
 
     if ($variables['theme_hook_original'] == 'select' && in_array($variables['element']['#id'], $slider_ids)) {
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: All this code could be split off in single functions implementing `hook_theme_suggestions_HOOK_alter()`. Only the floating part would stay in here (more or less).

File: web/themes/joinup/js/script.js [hunk 11]
```diff
@@ -3,22 +3,48 @@
...
+  Drupal.behaviors.silderSelect = {
+    attach: function (context, settings) {
+      $(context).find('.slider__select').once('sliderSelect').each(function () {
+         var select = $(this);
+         var selectLength = $(this).find('option').length;
 
-   $.each($('.slider__select'), function (index, value) {
-     var select = $(this);
-     var selectLength = $(this).find('option').length;
+         var slider = $("<div id='slider' class='slider__slider'></div>").insertAfter(select).slider({
```
Comment (brummbar) [inline web/themes/joinup/js/script.js]: Wrong tab size here (3 instead of 2).

File: web/themes/joinup/js/script.js [hunk 12]
```diff
@@ -3,22 +3,48 @@
...
+        $(this).addClass('button--default button--blue mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--accent');
+      });
+    }
+  };
+
+  Drupal.behaviors.silderSelect = {
+    attach: function (context, settings) {
+      $(context).find('.slider__select').once('sliderSelect').each(function () {
+         var select = $(this);
+         var selectLength = $(this).find('option').length;
```
Comment (brummbar) [inline web/themes/joinup/js/script.js]: Instead of `$(this)` use `select` variable to avoid creating again another jQuery object.
If possible and we all agree, I like more to put a dollar sign in front of variables that are jQuery objects. So it's easy to understand what is inside.

File: web/themes/joinup/js/script.js [hunk 13]
```diff
@@ -3,22 +3,48 @@
...
-        max: selectLength,
-        range: "min",
-        value: select[ 0 ].selectedIndex + 1,
-        change: function (event, ui) {
-          select.find('option').removeAttr('selected');
-          $(select.find('option')[ui.value - 1]).attr('selected', 'selected');
-        }
+  Drupal.behaviors.ajaxReload = {
+    attach: function (context, settings) {
+      $(document).ajaxComplete(function (event, xhr, settings) {
```
Comment (brummbar) [inline web/themes/joinup/js/script.js]: There is no `.once()` here so this event will be appended multiple times.

File: web/themes/joinup/templates/form/form-element--checkbox.html.twig [hunk 14]
```diff
@@ -82,13 +74,13 @@
       {{ description.content }}
     </div>
   {% endif %}
-  {{ children }}
+  <label class="mdl-checkbox mdl-js-checkbox mdl-js-ripple-effect">
```
Comment (brummbar) [inline web/themes/joinup/templates/form/form-element--checkbox.html.twig]: Why in the password twig file there is a checkbox class?

File: web/themes/joinup/templates/form/input--checkbox.html.twig [hunk 15]
```diff
@@ -10,4 +10,4 @@
  * @see template_preprocess_input()
  */
 #}
-<input{{ attributes.addClass('mdl-checkbox__input') }} />{{ children }}
+<input{{ attributes.addClass('mdl-checkbox__input') }} />
```
Comment (brummbar) [inline web/themes/joinup/templates/form/input--checkbox.html.twig]: Why {{ children }} has been left out?

File: web/themes/joinup/joinup_theme.theme (hunk 6 above)
Comment (tom-waat) [inline web/themes/joinup/joinup_theme.theme]: There is more than one parent for example in wysiwyg editor elements. "Text format" dropdown and "About text formats" link have also form_wrapper class, so I needed different class for form field.

File: web/themes/joinup/joinup_theme.theme (hunk 9 above)
Comment (tom-waat) [inline web/themes/joinup/joinup_theme.theme]: The #form_id is not accessible in variables in theme_suggestions_alter() function. But I think that we can use #id there (these elements are not duplicated or rebuilded by ajax).

File: web/themes/joinup/templates/form/form-element--checkbox.html.twig (hunk 14 above)
Comment (tom-waat) [inline web/themes/joinup/templates/form/form-element--checkbox.html.twig]: Hmm, I do not understand. It is not password template but form-element template.

File: web/themes/joinup/templates/form/input--checkbox.html.twig (hunk 15 above)
Comment (tom-waat) [inline web/themes/joinup/templates/form/input--checkbox.html.twig]: The children variable is empty for checkbox elements, so I think that it can be removed.

**Comment by tom-waat** [comment]
I fixed most fixed issues mentioned in the comments above. I added comments where something was unclear.

File: web/themes/joinup/templates/form/form-element--checkbox.html.twig (hunk 14 above)
Comment (brummbar) [inline web/themes/joinup/templates/form/form-element--checkbox.html.twig]: I think I was drunk. Ignore :)

File: web/themes/joinup/templates/form/input--checkbox.html.twig (hunk 15 above)
Comment (brummbar) [inline web/themes/joinup/templates/form/input--checkbox.html.twig]: If it doesn't create an issue, I would leave it there then, just to be sure.

File: web/themes/joinup/joinup_theme.theme (hunk 9 above)
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: I see that you removed this part of the code. Anyway, the `form_id` will always be there, it's set in the form builder. And the id will be modelled after it.

File: web/themes/joinup/templates/form/input--checkbox.html.twig (hunk 15 above)
Comment (tom-waat) [inline web/themes/joinup/templates/form/input--checkbox.html.twig]: Ok, I readded there {{ children }} variable.

File: web/themes/joinup/joinup_theme.theme [hunk 16]
```diff
@@ -101,28 +128,90 @@ function joinup_theme_preprocess_fieldset(&$variables) {
...
+    $suggestions[] = 'select__slider';
+  }
+}
+
+/**
+ * Implements hook_theme_suggestions_HOOK_alter() for input.html.twig.
+ */
+function joinup_theme_theme_suggestions_input_alter(array &$suggestions, array $variables) {
+  if ($variables['element']['#type'] == 'submit') {
+    $button_value = Html::cleanCssIdentifier($variables['element']['#value'], array(' ' => '_'));
```
Comment (brummbar) [inline web/themes/joinup/joinup_theme.theme]: Using the button value as theme suggestion is not really reliable, because in case of multilingual or label changes it's going to break. I don't see any templates committed for this suggestion, is it still needed?

Comment (tom-waat) [inline web/themes/joinup/joinup_theme.theme]: It was needed for submit buttons, but since I moved them to another ticket (https://webgate.ec.europa.eu/CITnet/jira/browse/ISAICP-2670), they are no longer needed here. I also noticed that during my latest merge, I should not have removed form suggestion function, it is needed in develop branch. I rebuilded this function like other suggestion functions.
